Firestore/Source/Core/FSTEventManager.mm

```diff
@@ -40,181 +44,38 @@
 
+namespace {
+
 /**
  * Holds the listeners and the last received ViewSnapshot for a query being tracked by
  * EventManager.
  */
-@interface FSTQueryListenersInfo : NSObject
-@property(nonatomic, assign, readwrite) TargetId targetID;
-@property(nonatomic, strong, readonly) NSMutableArray<FSTQueryListener *> *listeners;
-
-- (const absl::optional<ViewSnapshot> &)viewSnapshot;
-- (void)setViewSnapshot:(const absl::optional<ViewSnapshot> &)snapshot;
-
-@end
-
-@implementation FSTQueryListenersInfo {
-  absl::optional<ViewSnapshot> _viewSnapshot;
-}
-
-- (const absl::optional<ViewSnapshot> &)viewSnapshot {
-  return _viewSnapshot;
-}
-- (void)setViewSnapshot:(const absl::optional<ViewSnapshot> &)snapshot {
-  _viewSnapshot = snapshot;
-}
-
-- (instancetype)init {
-  if (self = [super init]) {
-    _listeners = [NSMutableArray array];
-  }
-  return self;
-}
-
-@end
-
-#pragma mark - FSTQueryListener
-
-@interface FSTQueryListener ()
-
-/** The last received view snapshot. */
-- (const absl::optional<ViewSnapshot> &)snapshot;
-
-/**
- * Initial snapshots (e.g. from cache) may not be propagated to the ViewSnapshotHandler.
- * This flag is set to YES once we've actually raised an event.
- */
-@property(nonatomic, assign, readwrite) BOOL raisedInitialEvent;
-
-/** The last online state this query listener got. */
-@property(nonatomic, assign, readwrite) OnlineState onlineState;
-
-@end
-
-@implementation FSTQueryListener {
-  ListenOptions _options;
-
-  absl::optional<ViewSnapshot> _snapshot;
-
-  /** The ViewSnapshotHandler associated with this query listener. */
-  ViewSnapshotHandler _viewSnapshotHandler;
-}
-
-- (instancetype)initWithQuery:(FSTQuery *)query
-                      options:(ListenOptions)options
-          viewSnapshotHandler:(ViewSnapshotHandler &&)viewSnapshotHandler {
-  if (self = [super init]) {
-    _query = query;
-    _options = std::move(options);
-    _viewSnapshotHandler = std::move(viewSnapshotHandler);
-    _raisedInitialEvent = NO;
-  }
-  return self;
-}
-
-- (const absl::optional<ViewSnapshot> &)snapshot {
-  return _snapshot;
-}
-
-- (void)queryDidChangeViewSnapshot:(ViewSnapshot)snapshot {
-  HARD_ASSERT(!snapshot.document_changes().empty() || snapshot.sync_state_changed(),
-              "We got a new snapshot with no changes?");
-
-  if (!_options.include_document_metadata_changes()) {
-    // Remove the metadata-only changes.
-    std::vector<DocumentViewChange> changes;
-    for (const DocumentViewChange &change : snapshot.document_changes()) {
-      if (change.type() != DocumentViewChange::Type::kMetadata) {
-        changes.push_back(change);
-      }
+struct QueryListenersInfo {
+  TargetId target_id;
+  std::vector<std::shared_ptr<QueryListener>> listeners;
+
+  void Erase(const std::shared_ptr<QueryListener> &listener) {
+    auto found = absl::c_find(listeners, listener);
+    HARD_ASSERT(found != listeners.end(), "Shouldn't remove listeners that don't exist");
+    if (found != listeners.end()) {
```

Comment:
Is this check necessary, seeing that we just asserted on the same condition above?